### ISSUE TITLE
[JavaScript] Type checking for function property behaves diferrently from method declaration syntax

### ISSUE THREAD

**Issue by xl1** [microsoft/TypeScript#42984]
# Bug Report
On JavaScript type checking (`checkJs` option), object initializer with function property (`{ foo: function () {} }`) emits ts2339 error unexpectedly, if that function gets and sets other properties of `this`.

### 🔎 Search Terms
checkJs function method definitions

### 🕗 Version & Regression Information
- v3 (at least v3.9.7 on playground) does not emit an error. v4 emits an error,

### ⏯ Playground Link
https://www.typescriptlang.org/play?declaration=false&useJavaScript=true#code/MYewdgzgLgBAHjAvDA3gKBpmBDAXDABgBoMsAjfY0zAMwFcxgoBLcAZQE8wps597GLcDAAUASlTUsmKAAtmEAHTYkMAIwBuKdNCQQAGwCmi-SADmIuQsVkxGmAHoHMAAoAnEAAdDbqBxgA5GQBMAAmIIYQMGAgsIZwCrDCft6BAkysYJzcvAGKIgBMAMxFAJxiUgC+JNIAtoZyIKHZPHDiktLSVkoqyJraWLoQBsamFt02dlU1WOlCWVytAOrMjXRQAIIQEMxmYPXc-AwZwu3onYPgw0Ym5pbySrZa0pVolRpAA

### 💻 Code

```js
// @ts-check
const x = {
    a: 0,
    b: 0,
    functionSyntax: function () {
        this.a = 1;
        console.log(this.b); // Property 'b' does not exist on type 'functionSyntax'.(2339)
    },
    methodSyntax() {
        this.a = 1;
        console.log(this.b); // no error
    },
    functionSyntaxWithoutAssignment: function () {
        console.log(this.b); // no error
    }
};
```

### 🙁 Actual behavior
ts(2339) error `Property '{name}' does not exist on type '{function name}'.` is reported

### 🙂 Expected behavior
No error

The method declaration syntax (`{ foo() {} }`) does not emit an error. I think that function syntax should have the same behavior for this checking.

**Comment by MartinJohns** [comment]
Intentional and mentioned in the FAQ. Check for bi-variance.

**Comment by xl1** [comment]
I don't think it is related to variance..., but I found a change that seemed to be relevant: https://github.com/microsoft/TypeScript/pull/39447 and understood its motivation.

With typescript v3, the code below generates an error. v4 is improved on this point.
```js
// @ts-check
const constructors = {
    SomeClass: function () {
        this.a = 1; // Property 'a' does not exist on type '{ SomeClass: typeof SomeClass; }'.(2339)
    },
};
const x = new constructors.SomeClass();
console.log(x.a);
```
In this situation it is impossible to distinguish between constructors and non-constructor functions, so the current behavior is certainly convincing.

I have a lot of code that uses non-constructor function properties, so I would be happy if there was any workaround for this error.

**Comment by xl1** [comment]
I will close this issue as I think this is an expected behavior.
